x-pack/plugins/cases/public/components/all_cases/use_all_cases_state.tsx

```diff
@@ -0,0 +1,224 @@
+    (params) => {
+      if (isModalView) {
+        setQueryParams((prevParams) => ({ ...prevParams, ...params }));
+        return;
+      }
+
+      const parsedUrlParams: ParsedUrlQueryParams = parse(location.search);
+      const urlParams: PartialQueryParams = parseUrlQueryParams(parsedUrlParams);
+      const newQueryParams: QueryParams = getQueryParams(
+        params,
+        queryParams,
+        urlParams,
+        localStorageQueryParams
+      );
+      const newLocalStorageQueryParams = {
+        perPage: newQueryParams.perPage,
+        sortField: newQueryParams.sortField,
+        sortOrder: newQueryParams.sortOrder,
+      };
+      setLocalStorageQueryParams(newLocalStorageQueryParams);
+      setQueryParams(newQueryParams);
+    },
+    [isModalView, location.search, queryParams, localStorageQueryParams, setLocalStorageQueryParams]
+  );
+
+  const persistAndUpdateFilterOptions = useCallback(
+    (params) => {
+      if (isModalView) {
+        setFilterOptions((prevParams) => ({ ...prevParams, ...params }));
+        return;
+      }
+
+      const newFilterOptions: FilterOptions = getFilterOptions(
+        filterOptions,
+        params,
+        parse(location.search),
+        localStorageFilterOptions
+      );
+
+      const newPersistedFilterOptions: PartialFilterOptions = {
+        ...(newFilterOptions.severity && { severity: newFilterOptions.severity }),
+        ...(newFilterOptions.status && { status: newFilterOptions.status }),
+      };
+
+      const newLocalStorageFilterOptions: PartialFilterOptions = {
+        ...localStorageFilterOptions,
+        ...newPersistedFilterOptions,
+      };
+      setLocalStorageFilterOptions(newLocalStorageFilterOptions);
+      setFilterOptions(newFilterOptions);
+    },
+    [
+      filterOptions,
+      isModalView,
+      localStorageFilterOptions,
+      location.search,
+      setLocalStorageFilterOptions,
+    ]
+  );
+
+  useEffect(() => {
+    if (isFirstRenderRef.current) {
+      persistAndUpdateQueryParams(isModalView ? queryParams : {});
+      persistAndUpdateFilterOptions(isModalView ? filterOptions : initialFilterOptions);
+
+      isFirstRenderRef.current = false;
+    }
+  }, [
+    filterOptions,
+    initialFilterOptions,
+    isModalView,
+    persistAndUpdateFilterOptions,
+    persistAndUpdateQueryParams,
+    queryParams,
+  ]);
+
+  if (!isModalView) {
+    const parsedUrlParams = parse(location.search);
+    const stateUrlParams = {
+      ...parsedUrlParams,
+      ...queryParams,
+      page: queryParams.page.toString(),
+      perPage: queryParams.perPage.toString(),
+      severity: filterOptions.severity,
+      status: filterOptions.status,
+    };
+
+    if (!isEqual(parsedUrlParams, stateUrlParams)) {
+      try {
+        const newHistory = {
+          ...location,
+          search: stringify({ ...parsedUrlParams, ...stateUrlParams }),
+        };
+
+        history.replace(newHistory);
```

Comment:
One of the warnings I was working on in a previous PR was the "state being updated during render" one. It was connected to `history.replace/history.push` not being called inside `useEffect`.

So initially when working on this I added a new `useEffect` hook just to update the history.

[But then I was thinking about this.](https://beta.reactjs.org/learn/you-might-not-need-an-effect#updating-state-based-on-props-or-state) And I thought it would be more elegant without the hook.

If the only consequence of the call is a test warning I don't mind it.

What do you think?

Reply:
Yes, tests are a good indicator that something is going wrong. No that you remove the `history` calls from the function s you can put the history in the `useEffect` but not use it for the state updates (`persistAndUpdateQueryParams` & `persistAndUpdateFilterOptions`)

Reply:
Done already 👌